framework/include/kernels/KernelScalarBase.h

```diff
@@ -0,0 +1,148 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "Kernel.h"
+
+/**
+ * This Kernel adds standardized methods for assembling to a primary
+ * scalar variable associated with the primary variable of the Kernel
+ * object. Essentially, the entire row of the residual and Jacobian
+ * associated with this scalar variable will also be assembled here
+ * using the loops over volumetric elements.
+ * This variable is "scalar_variable" in the input file and "kappa"
+ * within the source code.
+ */
+
+class KernelScalarBase : public Kernel
+{
+public:
+  static InputParameters validParams();
+
+  KernelScalarBase(const InputParameters & parameters);
+
+  virtual const MooseVariable & variable() const override { return _var; }
+
+  /**
+   * The scalar variable that this kernel operates on.
+   */
+  const MooseVariableScalar & scalarVariable() const
+  {
+    mooseAssert(_kappa_var_ptr, "kappa pointer should have been set in the constructor");
+    return *_kappa_var_ptr;
```

Comment:
An example of this use case is described near the bottom of `KernelScalarBase.md` in reference to the HomogenizedTotalLagrangianStressDivergenceR class. The current tensor mechanics homogenization system uses some kernels for the off-diagonal assembly and a scalar kernel for the scalar-scalar Jacobian and residual assembly. I have committed three versions of source files in the "test" directory to show how to reimplement that system all within one of these `KernelScalarBase` objects. However, since @reverendbedford has the homogenization kernel derived two or three levels deep off the Lagrangian classes, and I have to insert the `KernelScalarBase` between the `Kernel` and the first of those classes, I need to allow those intermediate classes NOT to have a scalar variable (e.g. `LagrangianStressDivergenceBaseS.C`).
Now, I'm new to C++, so maybe the polymorphic or the classes derived from multiple bases idea could solve the issue of not having intermediate classes which don't need the scalar variable. If it's always possible to do that multi-inheritance paths and not need the intermediate classes to derive off of `KernelScalarBase`, then we could eliminate the option for not using a scalar variable on a derived class.

Reply:
We can keep this for now and revise in the future (this is probably something I've also said before 😆 )